webextension/js/popup.js

```diff
@@ -1,150 +1,335 @@
+    console.log.call(console, ...args);
   }
-  document.querySelector(panelSelector).classList.remove("hide");
 }
 
-function showContainerTabsPanel(identity) {
-  // Populating the panel: name and icon
-  document.getElementById("container-info-name").innerText = identity.name;
+// This object controls all the panels, identities and many other things.
+let Logic = {
+  _identities: [],
+  _currentIdentity: null,
+  _currentPanel: null,
+  _previousPanel: null,
+  _panels: {},
+
+  init() {
+    // Retrieve the list of identities.
+    this.refreshIdentities()
+
+    // Routing to the correct panel.
+    .then(() => {
+      if (localStorage.getItem("onboarded2")) {
+        this.showPanel(P_CONTAINERS_LIST);
+      } else if (localStorage.getItem("onboarded1")) {
+        this.showPanel(P_ONBOARDING_2);
+      } else {
+        this.showPanel(P_ONBOARDING_1);
+      }
+    });
+  },
 
-  let icon = document.getElementById("container-info-icon");
-  icon.setAttribute("data-identity-icon", identity.image);
-  icon.setAttribute("data-identity-color", identity.color);
+  refreshIdentities() {
+    return browser.runtime.sendMessage({
+      method: "queryIdentities"
+    })
+    .then(identities => {
+      this._identities = identities;
+    });
+  },
 
-  // Show or not the has-tabs section.
-  for (let trHasTabs of document.getElementsByClassName("container-info-has-tabs")) {
-    trHasTabs.hidden = !identity.hasHiddenTabs && !identity.hasOpenTabs;
-    trHasTabs.setAttribute("data-user-context-id", identity.userContextId);
-  }
+  showPanel(panel, currentIdentity = null) {
+    // Invalid panel... ?!?
+    if (!(panel in this._panels)) {
+      throw("Something really bad happened. Unknown panel: " + panel + "\n");
+    }
 
-  let hideOrShowRow = document.querySelector("#container-info-hideorshow");
-  hideOrShowRow.setAttribute("data-user-context-id", identity.userContextId);
+    this._previousPanel = this._currentPanel;
+    this._currentPanel = panel;
 
-  const hideShowIcon = document.getElementById("container-info-hideorshow-icon");
-  hideShowIcon.src = identity.hasHiddenTabs ? CONTAINER_UNHIDE_SRC : CONTAINER_HIDE_SRC;
+    this._currentIdentity = currentIdentity;
 
-  const hideShowLabel = document.getElementById("container-info-hideorshow-label");
-  hideShowLabel.innerText = identity.hasHiddenTabs ? "Show these container tabs" : "Hide these container tabs";
+    // Initialize the panel before showing it.
+    this._panels[panel].prepare().then(() => {
+      for (let panelElement of document.querySelectorAll(".panel")) {
+        panelElement.classList.add("hide");
+      }
+      document.querySelector(this._panels[panel].panelSelector).classList.remove("hide");
+    });
+  },
 
-  // Let"s remove all the previous tabs.
-  for (const trTab of document.getElementsByClassName("container-info-tab")) {
-    trTab.remove();
-  }
+  showPreviousPanel() {
+    if (!this._previousPanel) {
+      throw "Current panel not set!";
```

Comment:
See previous comment re: `throw "Current panel not set!"` vs `throw new Error("Current panel not set!")`.

nit: Also, in some cases we're saying `throw "string"` and in other cases we're using `throw("string")` which made searching for instances a bit tricky.